libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestRequest.java

```diff
@@ -378,23 +473,152 @@ public static RestRequest getRequestForSearchResultLayout(String apiVersion, Lis
-		return new RestRequest(RestMethod.GET, path.toString(), null);	
-	}	
-	
-	/**
-	 * Jsonize map and create a RequestBody out of it
-	 * @param fields
+		return new RestRequest(RestMethod.GET, path.toString());
+	}
+
+    /**
+	 * Composite request
+	 * See https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_composite.htm
+	 *
+     * @param apiVersion
+     * @param allOrNone
+	 * @param refIdToRequests    Sorted map of reference id to refIdToRequests
 	 * @return
-	 * @throws UnsupportedEncodingException
 	 */
-	private static RequestBody prepareFieldsData(Map<String, Object> fields)
-			throws UnsupportedEncodingException {
-		if (fields == null) {
-			return null;
-		}
-		else {
-			return RequestBody.create(MEDIA_TYPE_JSON, new JSONObject(fields).toString());
+	public static RestRequest getCompositeRequest(String apiVersion, boolean allOrNone, SortedMap<String, RestRequest> refIdToRequests) throws JSONException {
+		JSONArray requestsArrayJson = new JSONArray();
+        for(Map.Entry<String,RestRequest> entry : refIdToRequests.entrySet()) {
+            String referenceId = entry.getKey();
+            RestRequest request = entry.getValue();
+            Map<String, String> headers = request.getAdditionalHttpHeaders();
+			JSONObject requestJson = new JSONObject();
+            requestJson.put(METHOD, request.getMethod().toString());
+            requestJson.put(URL, request.getPath());
+            requestJson.put(BODY, request.getRequestBodyAsJson());
+            if (headers != null) requestJson.put(HTTP_HEADERS, new JSONObject(headers));
+            requestJson.put(REFERENCE_ID, referenceId);
+			requestsArrayJson.put(requestJson);
 		}
+		JSONObject compositeRequestJson =  new JSONObject();
+		compositeRequestJson.put(COMPOSITE_REQUEST, requestsArrayJson);
+        compositeRequestJson.put(ALL_OR_NONE, allOrNone);
+
+		return new RestRequest(RestMethod.POST, RestAction.COMPOSITE.getPath(apiVersion), compositeRequestJson);
 	}
-	
+
+    /**
+     * Batch request
+     * See https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_batch.htm
+     *
+     * @param apiVersion
+     * @param haltOnError
+     *@param requests  @return
+     */
+    public static RestRequest getBatchRequest(String apiVersion, boolean haltOnError, RestRequest[] requests) throws JSONException {
+        JSONArray requestsArrayJson = new JSONArray();
+        for (RestRequest request : requests) {
+            // Note: unfortunately batch sub request and composite sub request differ
+            if (!request.getPath().startsWith(SERVICES_DATA)) {
+                throw new RuntimeException("Request not supported in batch: " + request.toString());
+            }
+            JSONObject requestJson = new JSONObject();
+            requestJson.put(METHOD, request.getMethod().toString());
+            requestJson.put(URL, request.getPath().substring(SERVICES_DATA.length()));
```

Comment:
Batch's sub request url is everything starts at the version (whereas composite sub request url is the whole path).